src/eval.ml

```diff
@@ -187,34 +196,55 @@ and tree_walk : dtree -> stack -> (term * stack) option = fun tree stk ->
                 env.(slot) <- TE_Some(b)
               else
                 let b = lift vars.(pos) in
-                let xs = Array.map (fun e -> VarMap.find e fresh_vars) xs in
+                let xs = Array.map (fun e -> IntMap.find e id_vars) xs in
                 env.(slot) <- TE_Some(Bindlib.unbox (Bindlib.bind_mvar xs b))
         in
         List.iter fn env_builder;
         (* Actually perform the action. *)
         Some(Bindlib.msubst act env, stk)
-    | Cond({ ok ; cond ; fail })                          ->
+    | Cond({ok; cond; fail})                              ->
         let next =
           match cond with
           | CondNL(i, j) ->
               if eq_modulo vars.(i) vars.(j) then ok else fail
           | CondFV(xs,i) ->
-              let xs = Array.map (fun e -> VarMap.find e fresh_vars) xs in
+              let allowed =
+                (* Term variables allowed in the term. *)
+                let _, dummy = IntMap.choose id_vars in
```

Comment:
What if id_vars is empty?

Reply:
Why not defining a dummy var independent of id_vars?

Reply:
id_vars shouldn't be empty since constraints concern variables that appear in the term.

Reply:
I suggest to simplify the code as follows:
```
let allowed_vars = Array.map (fun id -> try IntMap.find id id_vars with Not_found -> assert false) xs in
let other_vars = IntMap.fold (fun id var other -> if Array.mem id xs then other else var::other) id_vars [] in
let no_other_vars b = not (List.exists (fun v -> Bindlib.occur v b)) in
let b = Bindlib.bind_mvar allowed (lift vars.(i)) in
if no_other_vars b
then (bound.(i) <- TE_Some(Bindlib.unbox b); ok) else
...
```

Reply:
Here you traverse the [xs] array several times (with the [Array.mem]), whereas my code traverses it only once. My objective was to create the [allowed_vars] and the [other_vars] values in one traversal.

Reply:
Perhaps but [xs] is usually very short (size <= 1 in 99% of cases). So readability and conciseness is more important here.

Reply:
Fair enough.